Doc/library/sqlite3.rst

```diff
@@ -2614,6 +2595,77 @@ With some adjustments, the above recipe can be adapted to use a
 instead of a :class:`~collections.namedtuple`.
 
 
+.. _sqlite3-howto-text-factory:
+
+How to create and use text factories
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, :mod:`!sqlite3` adapts SQLite values with the ``TEXT`` data type
+using :class:`str`.
+This works well for correctly encoded UTF-8 text, but it will fail for invalid
+Unicode sequences and other encodings.
+To work around this, you can use a custom :attr:`~Connection.text_factory`.
+
+Because of SQLite's `flexible typing`_, it is not uncommon to encounter table
+columns with the ``TEXT`` data type, containing arbitrary data.
+To demonstrate, let's create a test database with an invalid Unicode sequence.
+We will use a `CAST expression`_ to coerce an invalid Unicode sequence,
+represented as a hexadecimal string ``X'619F'``,
+into the ``TEXT`` data type:
+
+.. testcode::
+
+   con = sqlite3.connect(":memory:")
+   con.executescript("""
+       CREATE TABLE test (data TEXT);
+       INSERT INTO test VALUES(CAST(X'619F' AS TEXT));
+   """)
+
+To work with such databases, we can use the following technique,
+borrowed from the :ref:`unicode-howto`:
+
+.. testcode::
+
+   con.text_factory = lambda data: str(data, errors="surrogateescape")
+   dump = con.iterdump()
+   for line in dump:
+       print(line)
+
+The loop above will print the offending line using Unicode surrogate escapes:
```

Comment:
"offending" might be a bit harsh if the DB simply uses a different encoding for all the text.  If instead a non-UTF-8 byte sequence managed to sneak in among the rest of the UTF-8 then the term might ok (even though I would be more offended by the code that would have let that happen in the first place).

Maybe it would be useful to address these two cases separately?

Reply:
> "offending" might be a bit harsh if the DB simply uses a different encoding for all the text.

+1! 😄 

> Maybe it would be useful to address these two cases separately?

Perhaps we only need one of them; I think the case of a different encoding makes for the better example.

Reply:
Maybe this is something that we could revisit in a separate PR, but I have the feeling that the focus of this howto is to explain how to use a certain API (namely creating/using text factories) rather than solving a real world problem (i.e. how to deal with non-UTF-8 text fields).  IOW, I think it's unlikely that users will come to the docs looking for text factories (i.e. a solution they might not be aware of) -- they will probably come with a decoding error that they can't figure out (i.e. a problem).

The decoding error might have happened for different reasons depending on what exactly it's on those fields, and the way you deal with it also changes:
* for fields that simply use a different encoding, the solution is to use a text factory that decodes the text using the correct encoding;
* for fields that should be using UTF-8 but contain some ill-encoded byte sequence, the solution is to use `surrogateescape` and friends to escape/ignore the invalid bytes.

Since both problems are related to encodings and can be solved with text factories, you could cover both in the same howto, by explaining the "different encoding" scenario and then just adding a paragraph at the end to mention how to use `surrogateescape`.

Reply:
I reworded the how-to to focus on the problem (encodings) instead of the solution (text factories). I simplified the text (how-to's are not explanations) and tried to cover both cases. How does 3b9f620aacac66c20344473a218af5d75959736f look, Ezio?